extensions/amp-smartlinks/0.1/linkmate.js

```diff
@@ -0,0 +1,180 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {deepEquals} from '../../../src/json';
+import {dict} from '../../../src/utils/object';
+
+import {ENDPOINTS} from './constants';
+import {TwoStepsResponse} from
+  '../../amp-skimlinks/0.1/link-rewriter/two-steps-response';
+import {getData} from '../../../src/event-helper';
+
+
+export class Linkmate {
+  /**
+   * @param {?../../../src/service/ampdoc-impl.AmpDoc} ampDoc
```

Comment:
I believe these params should be non nullable and it looks to be the case.
    
    @param {!....amp-doc-impl.AmpDoc} ampDoc
    @param {!....xhr-impl.Xhr} xhr
    @param {!Object} linkMateOptions

Otherwise instantiating the following fields in a Linkmate object would result in errors when trying to access a field in a possibly nullable linkmateOptions and ampDoc object.

    /** @private {?boolean} */
    this.requestExclusiveLinks_ = linkmateOptions.exclusiveLinks;

    /** @private {?number} */
    this.publisherID_ = linkmateOptions.publisherID;

    /** @private {?string} */
    this.linkAttribute_ = linkmateOptions.linkAttribute;
    
    /** @private {!Document|!ShadowRoot} */
    this.rootNode_ = this.ampDoc_.getRootNode();